asrtoolkit/alignment/align.py

```diff
@@ -0,0 +1,181 @@
+
+import io
+import json
+import logging
+import os
+import pickle
+from os.path import basename, dirname, exists
+from os.path import join as join_path
+from pathlib import Path
+
+import en_core_web_sm
+
+# Third Party
+import spacy
+from spacy.tokens import Doc as spacy_doc
+from textacy.extract import ngrams, noun_chunks
+from toolz.sandbox.core import pluck
+
+from asrtoolkit.alignment.align_utils import (
+    AlignedDoc,
+    Extractor,
+    dict_to_segments,
+    word_lattice_to_lines,
+)
+from asrtoolkit.alignment.initialize_logger import initialize_logger
+
+initialize_logger()
+LOGGER = logging.getLogger(__name__)
+NLP = en_core_web_sm.load()
+
+
+class WhitespaceTokenizer:
+    """
+    Text (List of Tokens) to Spacy Docs
+    Tokenizer input - list of words
+      returned spacy Doc tokens same as initial word list
+      equivalent to splitting on white space
+    """
+    def __init__(self, NLP):
+        """NLP: loaded Spacy model"""
+        self.vocab = NLP.vocab
+
+    def __call__(self, word_list):
+        """word_list: list of words; returns Doc"""
+        return spacy_doc(self.vocab,
+                         words=word_list,
+                         spaces=[True] * len(word_list))
+
+
+def init_spacy_document(lattice, token_key="token"):
+    """
+    Given List of word-level dicts, initializes Spacy Doc
+      using WhitespaceTokenizer
+    :param: lattice: List[Dict];
+      inner dicts contain token-level information for each token in input text
+    :param: token_key: key in Dicts containing tokens in text
+    :return: spacy Doc
+      (1) text: ' '.join(dict['token'])
+      (2) metadata: doc.user_data = lattice (word-level dicts)
+        - Scribe output: word-level time offsets
+        - reference transcript: speaker id and gender
+
+    * Note: spacy model must be loaded and bound to NLP object
+        NLP = spacy.load('en_core_web_sm')
+    """
+    NLP.make_doc = WhitespaceTokenizer(NLP)
+    word_list = list(pluck(token_key, lattice))
+    doc = NLP(word_list)
+    doc.user_data = lattice
+    assert len(doc) == len(word_list)  # 1:1 token:meta_dict
+    return doc
+
+
+def select_extractors(use_unigrams=False):
+    """
+    Extractors For Alignment
+    :return: List of Extractors objects to use for text-text alignment
+    note: ngram extractors below filter out stopwords and number words/symbols
+    """
+    noun_chunk_extractor = Extractor(
+        lambda doc: list(filter(lambda x: len(x) > 3, list(noun_chunks(doc)))))
+    tetragram_extractor = Extractor(
+        lambda doc: list(ngrams(doc, 4, filter_stops=True, filter_nums=True)))
+    trigram_extractor = Extractor(
+        lambda doc: list(ngrams(doc, 3, filter_stops=True, filter_nums=True)))
+    bigram_extractor = Extractor(lambda doc: list(
+        ngrams(doc, 2, filter_stops=False, filter_nums=False)))
+    unigram_extractor = Extractor(lambda doc: list(
+        ngrams(doc, 1, filter_stops=False, filter_nums=False)))
```

Comment:
What's the reason for not filtering stop words with the bigram and unigrams? Is that convention?

Reply:
I like the descriptive variable names for all these!

Reply:
I defer to @ageojo here.